### DISCUSSION TITLE
When to handle a keypress?

### DISCUSSION THREAD

**Discussion by jsholmes** [ArthurSonzogni/FTXUI#428] [Q&A]
Hi there.  I've got an app with a main component that contains a bunch of subcomponents.  The main component wants to display a help dialog when the user hits `h` or `?`.  This works great, but there are also some text input boxes where the user can legitimately type words that contain h, like `there`.  Right now as soon as the user hits the `h` in "there", the help dialog pops up.

How can the MainComponent know that some great-grandchild input component of it has Focus for input?

**Comment by ArthurSonzogni** [comment]
I think you could create a CatchEvent class that will first ask for the children.

This is similar to:
https://github.com/ArthurSonzogni/FTXUI/blob/master/src/ftxui/component/catch_event.cpp
but, we invert instructions in CatchEventBase2::OnEvent(...)

```cpp
class CatchEventBase2 : public ComponentBase {
 public:
  // Constructor.
  explicit CatchEventBase2(std::function<bool(Event)> on_event)
      : on_event_(std::move(on_event)) {}

  // Component implementation.
  bool OnEvent(Event event) override {
    if (ComponentBase2:OnEvent(event)) {
      return true;
    } else {
      return on_event_(event)
    }
  }

 protected:
  std::function<bool(Event)> on_event_;
};

Component CatchEvent2(Component child,
                     std::function<bool(Event event)> on_event) {
  auto out = Make<CatchEventBase2>(std::move(on_event));
  out->Add(std::move(child));
  return out;
}

ComponentDecorator CatchEvent2(std::function<bool(Event)> on_event) {
  return [on_event = std::move(on_event)](Component child) {
    return CatchEvent2(std::move(child), [on_event = on_event](Event event) {
      return on_event(std::move(event));
    });
  };
}
```

**Reply by jsholmes** [reply]
I wonder if maybe this is a better order to CatchEvent in by default?  It seems less surprising that if a child reports it handles the event, the parent won't have to process it.

But either way, I'll just add a policy to CatchEventBase that lets the user pick the order it catches in.  (Though I would suggest the default be the opposite for least surprise, but maybe I'm wrong?)

**Comment by jsholmes** [comment]
@ArthurSonzogni , a sort of related question:

Is it possible for a component to Unfocus/YieldFocus itself in its OnEvent handler?  For instance, I have a data table with rows.  The user can select rows with arrow keys or the mouse.  If they do, the table does TakeFocus().   If it is Focused(), then its Render() call, it shows a section with details about the row at the bottom.

When the user hits ESC, I'd like to stop showing the details at the bottom and in order to do that, unfocus the table.

**Reply by ArthurSonzogni** [reply]
What about calling "TakeFocus()" on the the component you want?

Only one can, so taking the focus somewhere else means giving back the focus from where it was originally.

**Reply by jsholmes** [reply]
I think I misunderstand Focus.

If a child of a component has focus, doesn't the component have focus, too?  I guess maybe I misunderstand focus.  The docs say:

```An element is Focused when it is with all its ancestors the ActiveChild() of their parents, and it is Focusable()```

If A is B's parent and B is focused, then that means B is the active child of A and Focusable.  But if A has some parent, P, isn't A the active child of P and Focusable?

In any event, what's the right way to tell if a child currently has focus?  Just iterate over the child hierarchy and check `Focused()` ?

**Reply by jsholmes** [reply]
Furthermore, if the A is the parent of B and B is the parent of C and C currently is Focused().

If either A or B calls TakeFocus(), I don't think this removes focus from C, since C remains the active child of B.

**Reply by ArthurSonzogni** [reply]
Component have children components. One of them is the "active child".

So ComponentBase::Active() is implemented by:
https://github.com/ArthurSonzogni/FTXUI/blob/ef0f1da147ca1a5579d2edecc4af29817e568c94/src/ftxui/component/component.cpp#L149-L151

If you start from the root and follow all the active child. You reached all the "Focused()" ones.
So Focused() is implemented by:
https://github.com/ArthurSonzogni/FTXUI/blob/ef0f1da147ca1a5579d2edecc4af29817e568c94/src/ftxui/component/component.cpp#L158-L164

Only one leaf component can be "Focused" at the same time. It represents where the users is on the screen.  When A component is Focused, all of its parents are focused.

**Reply by jsholmes** [reply]
Ah ok.  So to TakeFocus from one child, the parent has to focus some *other* leaf child.  There cannot be just no focused component once anything has ever been focused?

**Reply by jsholmes** [reply]
And in fact, because of the way _selector and selected are initialized, the leaf child along the 0th branch of every parent is Focused when the app starts, right?

**Reply by ArthurSonzogni** [reply]
Yes, the user needs to be somewhere.

Technically, you could create some container supporting being active, without any active child. The user would be "on" the container, without being on any specific children. Then you could goes into this state using: `Parent()->SetActiveChild(nullptr)`.

However, FTXUI do not support any component like this. We only expose "Container::Horizontal" & etc implemented by:
https://github.com/ArthurSonzogni/FTXUI/blob/ef0f1da147ca1a5579d2edecc4af29817e568c94/src/ftxui/component/container.cpp#L50-L57
https://github.com/ArthurSonzogni/FTXUI/blob/aebde943523bb75da79ba45eb5b2e88d39a620e1/src/ftxui/component/container.cpp#L42-L48
which do not support having no active children.

So, I would say no.

**Reply by jsholmes** [reply]
Ok.  I'll just define some ordering of the leaf components and make the 0th one the one I focus on when Esc is pressed.  Thanks for the help!
